### PR TITLE
WIP: Fixes empty segments creation in tsdb when presence of a very old block

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen Singh <harkishensingh@hotmail.com>

Fixes: #6359

Fixes empty segments creation when the most recent block is old. If no segments found, the `minTime` is shifted to the maxTime of the last block.

**Note:** Unit-test cases are not updated yet. Expect CI tests to fail.

### CODE REVIEW

**Comment by Harkishen-Singh** [comment]
cc @brian-brazil @codesome @krasi-georgiev

File: tsdb/db.go [hunk 1]
```diff
@@ -606,7 +607,29 @@ func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs
...
+		dec := record.Decoder{}
+		for reader.Next() {
+			currRecord := reader.Record()
+			if record.Samples == dec.Type(currRecord) {
+				var samples []record.RefSample
+				samples, err := dec.Samples(currRecord, samples)
+				if err != nil {
+					return nil, errors.Wrap(err, "samples append")
+				}
+				minValidTime = samples[0].T
```
Comment (brian-brazil) [inline tsdb/db.go:626]: The first sample is not required to be the one with the lowest timestamp.

Comment (Harkishen-Singh) [reply tsdb/db.go:626]: I think I did this since it was written in the issue as a desirable approach. Well, then should it be the minimum timestamp?

Comment (brian-brazil) [reply tsdb/db.go:626]: Yes, we need to consider all the samples.

**Comment by Harkishen-Singh** [comment]
@brian-brazil done!

**Comment by brian-brazil** [comment]
This would have us read the WAL twice at startup, which isn't desirable. The head rather than the db should also be dealing with WAL-level stuff.

**Comment by Harkishen-Singh** [comment]
@brian-brazil I hope the recent changes pushed is what we want.

@codesome, for the case involving no wal but a very old block, should we set the `h.minValidTime` to current timestamp?

**Comment by brian-brazil** [comment]
That looks better alright.

> for the case involving no wal but a very old block, should we set the h.minValidTime to current timestamp?

Could we fallback to whatever logic we use when there's no data at all?

**Comment by krasi-georgiev** [comment]
> @codesome, for the case involving no wal but a very old block, should we set the h.minValidTime to current timestamp?

I think to the maxt of the last block on disk.

**Comment by Harkishen-Singh** [comment]
How do we decided when to call a block very old. Like beyond a particular time range? 

Basically what I get through this discussion is that we simply consider the first sample. But, if wal is empty, take the maxt of last block. However, the later part will again cause lots of empty segments.

@krasi-georgiev @brian-brazil

**Comment by krasi-georgiev** [comment]
yeah I am not sure, I guess need to dig in the code and make a list of all cases.

NO WAL NO blocks.
NO WAL with blocks
With WAL and NO blocks
With WAL and with blocks

**Comment by brian-brazil** [comment]
We'd need to handle this at the compaction stage rather than the WAL reading stage to do that.

It also occurs to me that setting min valid time like this is incorrect, as it prevents ingesting data that is older than the oldest sample in the WAL. Not that that should be common, but it's not something we should break.

**Comment by krasi-georgiev** [comment]
@brian-brazil you probably have the most context here so can you sum up how should the head min/maxt be set then? 

Maybe a good start would be to remind us what was the accepted time range in the past that samples are accepted. For example if current timestamp is 10000 how much in the past would tsdb accept samples?

**Comment by brian-brazil** [comment]
1-3 hours, depending on where you are in the compaction cycle.

I suspect this is something we need to handle at the compaction stage.

**Comment by Harkishen-Singh** [comment]
@codesome as far as I remember, there was some other PR which covered this PR's purpose as well. Right?

**Comment by codesome** [comment]
That should be https://github.com/prometheus/prometheus/pull/7579

**Comment by roidelapluie** [comment]
We have looked at it in the bug scrub and decided to close this as superseded by  #7579

**Pull request closed without merging**